The goal of this notebook is to test the modularity data analysis pipeline on a test case. 

In [1]:
from neuprint import Client
# remove my token before making notebook public
c = Client('neuprint.janelia.org', dataset='hemibrain:v1.2.1', token='eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdnMjExNEBjb2x1bWJpYS5lZHUiLCJsZXZlbCI6Im5vYXV0aCIsImltYWdlLXVybCI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hLS9BT2gxNEdpb1lJLUVPLWdidGxPRTh6SmQ0eF9ZQ1Y4ZHF0YVFjWGlHeG5CMz1zOTYtYz9zej01MD9zej01MCIsImV4cCI6MTgxMDUyOTYzNH0.jv9eR0SH5RhfBdXrtp4r-dDFOhcsT8GBbE4v69ysCKs') 
c.fetch_version()

# import important stuff here
import numpy as np
import pandas as pd
import matplotlib

In [2]:
# testing on a neuron with few connections
test_neuron_Id = 676124666# 1815929980 #450971893 #423101189 #

In [3]:
from neuprint import fetch_neurons
test_neuron_df, test_syns = fetch_neurons(test_neuron_Id)
test_neuron_df

bodyId instance  type  pre  post  downstream  upstream  mito    size  \
0  676124666     None  None    1     1           9         1     0  178283   

  status cropped statusLabel cellBodyFiber somaRadius somaLocation  \
0   None    None        None          None       None         None   

                                             roiInfo notes         inputRois  \
0  {'SNP(R)': {'pre': 1, 'post': 1, 'downstream':...  None  [SMP(R), SNP(R)]   

         outputRois  
0  [SMP(R), SNP(R)]

In [4]:
test_syns

bodyId         roi  pre  post  downstream  upstream  mito
0  676124666  NotPrimary    0     0           0         0     0
1  676124666      SMP(R)    1     1           9         1     0
2  676124666      SNP(R)    1     1           9         1     0

In [5]:
from neuprint import fetch_simple_connections

test_inputs = fetch_simple_connections(None, test_neuron_Id)
test_inputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0   644421171    676124666       1     None      None         None   

  instance_post                                       conn_roiInfo  
0          None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...

All of these downstream neurons are contacted from the same pre site. Interestingly, it seems that weight is not necessarily a count of the post sites as I assumed since the weight would've added up to the downstream total. This is true regardless of whether fetch_simple_connections or fetch_adjacencies is used. This is important to note for other stuff. Even when include_nonprimary=True for fetch_adjacencies, there are only 8 entries and it's just because synapses are being double counted in SMP and SNP rois.

In [6]:
from neuprint import fetch_simple_connections

test_outputs = fetch_simple_connections(test_neuron_Id,None)
test_outputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0   676124666    423101189       1     None     oviIN         None   
1   676124666    644761952       1     None   LHPD2c5         None   
2   676124666   5813009620       1     None    SMP109         None   
3   676124666   5813056054       1     None    SMP050         None   

  instance_post                                       conn_roiInfo  
0       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
1     LHPD2c5_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2      SMP109_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3      SMP050_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...

In [7]:
from neuprint import fetch_adjacencies

[testn_, test_outfetch] = fetch_adjacencies(test_neuron_Id,None)
test_outfetch

bodyId_pre  bodyId_post     roi  weight
0   676124666    423101189  SMP(R)       1
1   676124666    644761952  SMP(R)       1
2   676124666   5813009620  SMP(R)       1
3   676124666   5813056054  SMP(R)       1

In [8]:
from neuprint import fetch_adjacencies

[testn_, test_outfetch] = fetch_adjacencies(test_neuron_Id,None, include_nonprimary=True)
test_outfetch

bodyId_pre  bodyId_post     roi  weight
0   676124666    423101189  SNP(R)       1
1   676124666    423101189  SMP(R)       1
2   676124666    644761952  SNP(R)       1
3   676124666    644761952  SMP(R)       1
4   676124666   5813009620  SNP(R)       1
5   676124666   5813009620  SMP(R)       1
6   676124666   5813056054  SNP(R)       1
7   676124666   5813056054  SMP(R)       1

Just to harp on this point, this particular neuron (id: 676124666) has 1 pre site and 1 post site on its body. Each of its 4 downstream partners is connecting to the same pre site. You can see this when you fetch_synapse_connections for this neuron since all the pre sites have the exact same coordinates. It is entirely unclear what the "downstream" count of 9 is referring to. I will have to revisit anywhere we have used the downstream and upstream information and double check those results.

In [9]:
from neuprint import fetch_synapses

fetch_synapses(test_neuron_Id)

bodyId  type     roi      x      y     z  confidence
0  676124666   pre  SMP(R)  18027  31040  9294    0.902000
1  676124666  post  SMP(R)  18058  31070  9291    0.677568

In [10]:
from neuprint import fetch_synapse_connections
fetch_synapse_connections(test_neuron_Id,None)

  0%|          | 0/4 [00:00<?, ?it/s]

bodyId_pre  bodyId_post roi_pre roi_post  x_pre  y_pre  z_pre  x_post  \
0   676124666   5813056054  SMP(R)   SMP(R)  18027  31040   9294   17990   
1   676124666    423101189  SMP(R)   SMP(R)  18027  31040   9294   18025   
2   676124666    644761952  SMP(R)   SMP(R)  18027  31040   9294   18039   
3   676124666   5813009620  SMP(R)   SMP(R)  18027  31040   9294   18030   

   y_post  z_post  confidence_pre  confidence_post  
0   31030    9254           0.902            0.936  
1   31034    9274           0.902            0.994  
2   31043    9274           0.902            0.995  
3   31020    9323           0.902            0.370

To move on from this issue, I next test out the different options for the get_connectome function. Below is the default.
# Testing get_connectome

In [11]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171    676124666       1
7    644421171   5813009620       1
8    644421171   5813056054       1
9    644761952    423101189      20
10   644761952   5813009620      18
11   644761952   5813056054       5
12   676124666    423101189       1
13   676124666    644761952       1
14   676124666   5813009620       1
15   676124666   5813056054       1
16  5813009620    423101189      92
17  5813009620   5813056054      12
18  5813056054    423101189       2
19  5813056054   5813009620       1

This was done in the function by taking ...

If we exclude the main neuron, you can count that 5 rows from above should be excluded and this is indeed the result below.

In [12]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, exclude_main_neurons=True)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171   5813009620       1
7    644421171   5813056054       1
8    644761952    423101189      20
9    644761952   5813009620      18
10   644761952   5813056054       5
11  5813009620    423101189      92
12  5813009620   5813056054      12
13  5813056054    423101189       2
14  5813056054   5813009620       1

If we change the scope of the connectome to be inputs only, we get the only input connection which is with the main neuron itself.

In [14]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_scope='input')
test_connectome

bodyId_pre  bodyId_post  weight
0   644421171    676124666       1

Conversely, changing the scope to output only excludes the 6 rows where that only input neuron (id: 644421171) appeared.

In [15]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_scope='output')
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644761952      59
1    423101189   5813009620      15
2    423101189   5813056054      78
3    644761952    423101189      20
4    644761952   5813009620      18
5    644761952   5813056054       5
6    676124666    423101189       1
7    676124666    644761952       1
8    676124666   5813009620       1
9    676124666   5813056054       1
10  5813009620    423101189      92
11  5813009620   5813056054      12
12  5813056054    423101189       2
13  5813056054   5813009620       1

To test the thresholding by weight and the connectome by type options, we'll need a different test case that has multiple instances of those things.

In [24]:
# testing on a neuron with few connections
test_neuron_Id = 5813077939

In [25]:
from neuprint import fetch_neurons
test_neuron_df, test_syns = fetch_neurons(test_neuron_Id)
test_neuron_df

bodyId            instance  type  pre  post  downstream  upstream  \
0  5813077939  (aDT6a)(MBDLaxon2)  None   88   108         734       108   

   mito       size  status  cropped     statusLabel cellBodyFiber somaRadius  \
0    41  115647744  Traced    False  Roughly traced          None       None   

  somaLocation                                            roiInfo notes  \
0         None  {'SNP(R)': {'pre': 40, 'post': 49, 'downstream...  None   

                          inputRois                        outputRois  
0  [SMP(L), SMP(R), SNP(L), SNP(R)]  [SMP(L), SMP(R), SNP(L), SNP(R)]

In [26]:
from neuprint import fetch_simple_connections

test_inputs = fetch_simple_connections(None, test_neuron_Id)
test_inputs

bodyId_pre  bodyId_post  weight type_pre type_post        instance_pre  \
0   5901203505   5813077939      11   SLP212      None            SLP212_R   
1    550301858   5813077939       9     None      None                None   
2    516943720   5813077939       6     None      None  (aDT6f)(MBDLaxon2)   
3   1141035086   5813077939       5     None      None  (aDT6f)(MBDLaxon2)   
4    486544396   5813077939       4    PAL02      None             PAL02_R   
5    549978347   5813077939       4     None      None         (MBDLaxon2)   
6    704487840   5813077939       4   SMP602      None    SMP602(SCB061)_R   
7    392325925   5813077939       3   SMP103      None            SMP103_R   
8    455500416   5813077939       3    PAL02      None      PAL02(ADL27)_L   
9    613822435   5813077939       3   SMP103      None            SMP103_R   
10  5813012979   5813077939       3     None      None         (MBDLaxon2)   
11  5813055773   5813077939       3   SMP094      None            SMP094_R   
12  5901207157   5813077939       3     None      None           (ADM04)_L   
13   361105354   5813077939       2     None      None  (aDT6f)(MBDLaxon2)   
14   516283789   5813077939       2   SMP589      None     SMP589(PVM12)_L   
15   547612423   5813077939       2     None      None           (ADM04)_L   
16   644778919   5813077939       2     None      None                None   
17   674152184   5813077939       2     None      None  (aDT6a)(MBDLaxon2)   
18  5813012274   5813077939       2     None      None         (MBDLaxon2)   
19  5813047306   5813077939       2     None      None         (MBDLaxon2)   
20  5813063587   5813077939       2     pC1d      None              pC1d_R   
21   423442357   5813077939       1   SMP480      None            SMP480_R   
22   518934633   5813077939       1     None      None  (aDT6b)(MBDLaxon2)   
23   519603027   5813077939       1     None      None                None   
24   548660840   5813077939       1     None      None  (aDT6f)(MBDLaxon2)   
25   549364820   5813077939       1     None      None                None   
26   550388353   5813077939       1     None      None  (aDT6a)(MBDLaxon2)   
27   578642656   5813077939       1     None      None           (ADM04)_L   
28   582027317   5813077939       1   aSP-f2      None            aSP-f2_R   
29   582700300   5813077939       1   SMP004      None            SMP004_R   
30   612099662   5813077939       1     None      None         (MBDLaxon2)   
31   643462563   5813077939       1     None      None                None   
32   645556898   5813077939       1     None      None                None   
33   674480029   5813077939       1     None      None         (MBDLaxon2)   
34   674502264   5813077939       1     None      None  (aDT6a)(MBDLaxon2)   
35   675183666   5813077939       1     None      None                None   
36  5812980529   5813077939       1   SLP212      None            SLP212_R   
37  5813039948   5813077939       1     None      None  (aDT6b)(MBDLaxon2)   
38  5813069630   5813077939       1  NPFL1-I      None           NPFL1-I_R   
39  5813133687   5813077939       1     None      None                None   

         instance_post                                       conn_roiInfo  
0   (aDT6a)(MBDLaxon2)  {'SNP(R)': {'pre': 5, 'post': 5}, 'SMP(R)': {'...  
1   (aDT6a)(MBDLaxon2)  {'SNP(R)': {'pre': 7, 'post': 7}, 'SMP(R)': {'...  
2   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 6, 'post': 6}, 'SMP(L)': {'...  
3   (aDT6a)(MBDLaxon2)  {'SNP(R)': {'pre': 4, 'post': 4}, 'SMP(R)': {'...  
4   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 2, 'post': 2}, 'SMP(L)': {'...  
5   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 2, 'post': 2}, 'SMP(L)': {'...  
6   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 4, 'post': 4}, 'SMP(L)': {'...  
7   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 2, 'post': 2}, 'SMP(L)': {'...  
8   (aDT6a)(MBDLaxon2)  {'SNP(L)': {'pre': 2, 'post': 2}, 'SMP(L)': {'...  
9   (aDT6a)(MBDLaxon2)  {'SNP(R)': {'pre': 1,

In [27]:
from neuprint import fetch_simple_connections

test_outputs = fetch_simple_connections(test_neuron_Id,None)
test_outputs

bodyId_pre  bodyId_post  weight type_pre type_post        instance_pre  \
0   5813077939    579700707      23     None    SMP550  (aDT6a)(MBDLaxon2)   
1   5813077939    452689494      17     None    SMP550  (aDT6a)(MBDLaxon2)   
2   5813077939    456523994       9     None    SMP165  (aDT6a)(MBDLaxon2)   
3   5813077939    486544396       9     None     PAL02  (aDT6a)(MBDLaxon2)   
4   5813077939    642763374       8     None    oviDNb  (aDT6a)(MBDLaxon2)   
..         ...          ...     ...      ...       ...                 ...   
92  5813077939   5813019681       1     None    SMP078  (aDT6a)(MBDLaxon2)   
93  5813077939   5813055773       1     None    SMP094  (aDT6a)(MBDLaxon2)   
94  5813077939   5813060687       1     None      None  (aDT6a)(MBDLaxon2)   
95  5813077939   5813082394       1     None      None  (aDT6a)(MBDLaxon2)   
96  5813077939   5813134052       1     None    SMP116  (aDT6a)(MBDLaxon2)   

      instance_post                                       conn_roiInfo  
0   SMP550(PDM29)_L  {'SNP(L)': {'pre': 22, 'post': 22}, 'SMP(L)': ...  
1          SMP550_R  {'SNP(R)': {'pre': 17, 'post': 17}, 'SMP(R)': ...  
2          SMP165_R  {'SNP(R)': {'pre': 9, 'post': 9}, 'SMP(R)': {'...  
3           PAL02_R  {'SNP(R)': {'pre': 7, 'post': 7}, 'SMP(R)': {'...  
4   oviDNb(PDM15)_L  {'SNP(R)': {'pre': 7, 'post': 7}, 'SMP(R)': {'...  
..              ...                                                ...  
92  SMP078(ADM04)_L  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
93         SMP094_R  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
94      (MBDLaxon2)  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
95             None  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
96         SMP116_L  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  

[97 rows x 8 columns]

In [29]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id)
test_connectome

bodyId_pre  bodyId_post  weight
0      267214250    297921608     117
1      267214250    298258513       8
2      267214250    298957166       1
3      267214250    329728593       1
4      267214250    330661529      17
...          ...          ...     ...
3308  5901207157   5813055769       1
3309  5901207157   5813055773       2
3310  5901207157   5813069072       4
3311  5901207157   5813069630       1
3312  5901207157   5813077939       3

[3313 rows x 3 columns]

In [ ]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_by_type=True)
test_connectome

type_pre type_post  weight
0     CRE005    CRE044       4
1     CRE005    CRE107      92
2     CRE005    PPL102       1
3     CRE005    SMP078       5
4     CRE005    SMP109      12
..       ...       ...     ...
632     pC1e    SMP555      41
633     pC1e    SMP556      49
634     pC1e    SMP589       5
635     pC1e      pC1b       2
636     pC1e      pC1d       8

[637 rows x 3 columns]

In [31]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id, connectome_by_type=True, weight_threshold=100)
test_connectome

type_pre type_post  weight
0   SMP078    SMP108     116
1   SMP108    SMP078     110
2   SMP108    SMP116     130
3   SMP109    SMP116     110
4   SMP333    SMP108     133
5   SMP556    SMP109     100
6     pC1b    SMP333     117
7     pC1d    SMP555     185
8     pC1d    SMP556     339

# Official start of analysis pipeline
Now that I got to see exactly what this test neuron is connected to, the official analysis pipeline starts below. 

I use the get_connectome function to get a dataframe with the neuron's connectome. I want connectome_to_undirected to be a function that can be called separately and that returns the directed connectome dataframe. It is useful for these to be separate so that we can do other analyses with them besides modularity. 

In [7]:
# obtain the connectome dataframe for the test neuron
from get_connectome import get_connectome

test_connectome = get_connectome(test_neuron_Id)
test_connectome

bodyId_pre  bodyId_post  weight
0    423101189    644421171       9
1    423101189    644761952      59
2    423101189   5813009620      15
3    423101189   5813056054      78
4    644421171    423101189      51
5    644421171    644761952       3
6    644421171    676124666       1
7    644421171   5813009620       1
8    644421171   5813056054       1
9    644761952    423101189      20
10   644761952   5813009620      18
11   644761952   5813056054       5
12   676124666    423101189       1
13   676124666    644761952       1
14   676124666   5813009620       1
15   676124666   5813056054       1
16  5813009620    423101189      92
17  5813009620   5813056054      12
18  5813056054    423101189       2
19  5813056054   5813009620       1

In [8]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

test_connectome_undirected = connectome_to_undirected(test_connectome)
test_connectome_undirected

source      target  weight
0    423101189   644421171      60
1    423101189   644761952      79
2    423101189  5813009620     107
3    423101189  5813056054      80
4    644421171   644761952       3
5    644421171   676124666       1
6    644421171  5813009620       1
7    644421171  5813056054       1
8    644761952  5813009620      18
9    644761952  5813056054       5
10   676124666   423101189       1
11   676124666   644761952       1
12   676124666  5813009620       1
13   676124666  5813056054       1
14  5813009620  5813056054      13

The undirected connectome dataframe is ready to be exported to a csv file. From there it will be formatted using format_edgelist.py from the terminal window so that the nodes are no longer bodyIds but sequential numbers instead. Then the modularity should be able to run. It turns out that Alex's newest version of format_edgelist.py also makes the connectome undirected, so running the connectome_to_undirected function before format_edgelist.py is not necessary but also seems to be harmless. I verified that the results are the same regardless of whether connectome_to_undirected is run or not.

When exporting, either remove the header row or make sure to use the header flag when calling format_edgelist.py. Appropriate results were obtained with
`python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window using the exported file created below. test_connectome_undirected.txt gets overwritten and replaced by the sequentially numbered nodes. A copy of the original is created that is called original_test_connectome_undirected.txt, as well as a few other files (clean_, degree_, info_, and key_ files).

The exported file can be .txt or .csv. The files that are created from format_edgelist.py will inherit the same file type. I chose to export as .txt.

In [12]:
# export the undirected connectome to a csv or txt file
test_connectome_undirected.to_csv('test_connectome_undirected.txt', index = False, header=False)
#test_connectome_undirected.to_csv('test_connectome_undirected.txt', index = False, header=False, sep=' ')

After running format_edgelist.py in the terminal window, I then run: `sh work.sh test_connectome_undirected.txt` but I think this actually generates the same files that Alex generates with his new format_edgelist.py. The only problem with going directly to work.sh after saving test_connectome_undirected.txt is that work.sh doesn't renumber the nodes which we need. So effectively, format_edgelist.py rolls that step into the work.sh step all in one go. Alternatively, we could just run the functions separately from format_edgelist.py in this notebook. To do that, I had to change the separator to ' ' when exporting 'test_connectome_undirected.txt' in order to avoid errors. Below I made an attempt but the resulting new_test_connectome_undirected.txt file had renumbered the nodes correctly but the numbers were not sequential which was odd. I think it would be better to run `format_edgelist(input)` in a notebook where the input is a struct that contains prefix, suffix, and file path as far as I can tell. This would run all the things the way that running format_edgelist.py in the terminal does. I can play around with that if needed but I'll move on since we have a working solution which is to simply run format_edgelist.py in the terminal.

In [ ]:
# alternative to running format_edgelist.py from command line in terminal window
from format_edgelist import read_graph
from format_edgelist import write_edges
from pathlib import Path

myfile = Path('test_connectome_undirected.txt')
nodes, degrees, edges = read_graph(myfile)

myoutfile = Path('new_test_connectome_undirected.txt')
write_edges(myoutfile, edges, nodes)

If I run `python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window and follow that with `clang main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm` and `./a.out 2 5 2 12345 0 test_connectome_undirected.txt`, the results_ file is different than when I include the `sh work.sh test_connectome_undirected.txt` step. In the test that I did, the results ended up with more communities and slightly different numbers of neurons in them. It is unclear to me which is more correct. If I run the `sh work.sh test_connectome_undirected.txt` step twice, I don't get different results from when I run it once, so that makes me think that the reason results are different when I don't run it is because it does something that `python format_edgelist.py test_connectome_undirected.txt --sep comma` doesn't do. So I will continue including the `sh work.sh test_connectome_undirected.txt` step even though it is partially redundant with some of what format_edgelist.py does.

The pipeline so far is:

First run `python format_edgelist.py test_connectome_undirected.txt --sep comma` in the terminal window. test_connectome_undirected.txt is overwritten with an undirected edge list that contains sequentially numbered nodes. 5 new files appear: original_, clean_, key_, info_, degree_.

Then run `sh work.sh test_connectome_undirected.txt`. After running, the required files appear (if format_edgelist.py wasn't used to create them) and we next compile the C codes. I had to modify Pramesh's command to use clang given the weird compatibility issues with new macs and gcc and CMake.

`clang main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm` to generate the a.out file. On my computer I get a bunch of warnings but none seems important.


`./a.out 2 5 2 12345 0 test_connectome_undirected.txt` is supposed to generate results_test_connectome_undirected.txt and partition_test_connectome_undirected.txt, although in these test cases that partition_ file didn't appear. In fact, it seems like the C program doesn't finish running even when I use a tiny connectome.

When I tested this out on oviIN_R's full connectome with chi=0 I thought I ended up with 181 clusters but it turns out, those numbers are iterations of the algorithm. It seems like the algorithm doesn't finish running on my computer for some reason because it is supposed to end with a final Q value. 

I might be using improper parameters for the CPUs, etc but Rhessa uses the same default CPU parameters and even the same random seed. I suspect that something weird is going on with a library somewhere since the same thing was happening on the lab computer and even compiling the codes was giving me problems. 

Sidenote: When trying to do this on the lab computer, I had to compile the code using a bunch of additional flags so that the libomp library could be accessed since it's not in the place where clang expects: `clang -I/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/clang/14.0.6/include/ -L/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/  main.c help.c rg.c -Xpreprocessor -fopenmp -lomp -lm -Wl,-rpath,/Users/ggutierr/miniconda3/pkgs/llvm-openmp-14.0.6-hc6e5704_0/lib/`
However, the ./a.out step isn't working and it is probably for the same reason but I haven't figured out a work around. I might just have to install Homebrew and then install libomp with it so that the libraries are where they should be. Or I can write a Make file to point things to the right directories and libraries.

To read the modularity data back into Python, we can use preprocess.py from Alex's repo. Or Rhessa made a script called process.py that might do this.
Rhessa first combined modularity data into an xls spreadsheet and then runs process.py in terminal on that and other modularity files.

It might not be necessary to run this script if we just import the data from a csv into a dataframe, but I'm not sure.